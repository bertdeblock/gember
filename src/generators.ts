import { generate } from "./generate.js";

export function generateComponent(
  name: string,
  packagePath: string,
  {
    classBased = false,
    nested,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    nested?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate("component", name, packagePath, {
    inputs: { classBased, typescript },
    nested,
    path,
  });
}

export function generateHelper(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate("helper", name, packagePath, {
    inputs: { classBased, typescript },
    path,
  });
}

export function generateModifier(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate("modifier", name, packagePath, {
    inputs: { classBased, typescript },
    path,
  });
}

export function generateService(
  name: string,
  packagePath: string,
  {
    path,
    typescript = false,
  }: {
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate("service", name, packagePath, {
    inputs: { typescript },
    path,
  });
}
