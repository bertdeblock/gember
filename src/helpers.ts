import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
