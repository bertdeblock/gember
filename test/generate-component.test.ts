import fsExtra from "fs-extra";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { afterEach, it } from "vitest";
import { generateComponent } from "../src/generators.ts";
import { copyBlueprint } from "./helpers.ts";

let cwd: string;

afterEach(() => fsExtra.remove(cwd));

it("generates a template-only `.gjs` component", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { cwd });

  const content = await readFile(join(cwd, "src/components/foo.gjs"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.gjs` component", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { classBased: true, cwd });

  const content = await readFile(join(cwd, "src/components/foo.gjs"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a template-only `.gjs` component at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { cwd, path: "src/-private" });

  const content = await readFile(join(cwd, "src/-private/foo.gjs"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a template-only `.gts` component", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { cwd, ts: true });

  const content = await readFile(join(cwd, "src/components/foo.gts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.gts` component", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { classBased: true, cwd, ts: true });

  const content = await readFile(join(cwd, "src/components/foo.gts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a template-only `.gts` component at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateComponent("foo", { cwd, path: "src/-private", ts: true });

  const content = await readFile(join(cwd, "src/-private/foo.gts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});
