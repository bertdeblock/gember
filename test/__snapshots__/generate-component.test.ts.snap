// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a class-based \`.gjs\` component 1`] = `
"import Component from "@glimmer/component";

export default class Foo extends Component {
  <template>{{yield}}</template>
}
"
`;

exports[`generates a class-based \`.gts\` component 1`] = `
"import Component from "@glimmer/component";

export interface FooSignature {
  Args: {};
  Blocks: {
    default: [];
  };
  Element: null;
}

export default class Foo extends Component<FooSignature> {
  <template>
    {{yield}}
  </template>
}
"
`;

exports[`generates a nested colocated template-only \`.gjs\` component 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`generates a nested template-only \`.gjs\` component 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`generates a template-only \`.gjs\` component 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`generates a template-only \`.gjs\` component at a custom path 1`] = `
"<template>{{yield}}</template>
"
`;

exports[`generates a template-only \`.gts\` component 1`] = `
"import type { TOC } from '@ember/component/template-only';

export interface FooSignature {
  Args: {};
  Blocks: {
    default: [];
  };
  Element: null;
}

const Foo: TOC<FooSignature> = <template>{{yield}}</template>;

export default Foo;
"
`;

exports[`generates a template-only \`.gts\` component at a custom path 1`] = `
"import type { TOC } from '@ember/component/template-only';

export interface FooSignature {
  Args: {};
  Blocks: {
    default: [];
  };
  Element: null;
}

const Foo: TOC<FooSignature> = <template>{{yield}}</template>;

export default Foo;
"
`;
